Store initial blockchain state in TestAPIManager

AndrejMitrovic opened the pull request:
This will enable dependency injection for #785.

AndrejMitrovic committed:
Store the blocks in TestAPIManager

The unittests need to have access to the
Genesis block, which could be a custom
block or one which was automatically generated
in the Base.d module.

This will be required for #785

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -492,6 +498,34 @@ public interface TestAPI : ValidatorAPI
     public abstract void broadcastPreimage (uint height);
 }
 
+/*******************************************************************************
+
+    For dependency injection.
+
+*******************************************************************************/
+
+public class PreLoadedMemBlockStorage : MemBlockStorage
+{
+    /// Preloaded blocks, including the genesis block
+    immutable(Block)[] blocks;
+
+    /// Ctor
+    public this (immutable(Block)[] blocks)
+    {
+        this.blocks = blocks;
+    }
+
+    /// Load the blockchain
+    public override bool load ()
+    {
+        foreach (const ref block; this.blocks)
+            this.saveBlock(block);
+
+        return true;
+    }
+}
```
Does it makes sense to have it here ? That could just replace the MemBlockStorage, couldn't it ? Do you remember why `MemBlockStorage` has serialized Blocks ?

AndrejMitrovic replied:
> Does it makes sense to have it here ? 

Ah it should probably not be `public`. I can move it inside of `TestNodeMixin` I suppose. I only have one use-case for it and it's here in the tests.

> That could just replace the MemBlockStorage, couldn't it ? Do you remember why MemBlockStorage has serialized Blocks ?

Not sure. I think we can just remove that part? Then I could just add a constructor to `MemBlockStorage` and get rid of `PreLoadedMemBlockStorage` I suppose.

AndrejMitrovic commented:
Ok I see why it's using serialization.

Because it needs to store `const(Block)` in `saveBlock`, but it needs to load mutable `Block` in e.g. `tryReadBlock`.

AndrejMitrovic commented:
I've removed the extra class and just moved this stuff into the `MemBlockStorage` constructor.

Geod24 commented on source/agora/node/BlockStorage.d:
```diff
@@ -918,17 +918,23 @@ public class MemBlockStorage : IBlockStorage
     public override bool load ()
     {
-        if (this.blocks.length == 0)
-            this.saveBlock(GenesisBlock);
```
I would add a boolean in the new class, or at least something that checks we don't accidentally forget to initialize the BlockStorage from our code. [It happened before](https://github.com/bpfkorea/agora/commit/c7b90eb9483607d900fef63c3a7b63b2cae2b5dc)

AndrejMitrovic replied:
Ah then I'll move the loading into this method.

Geod24 commented:
[This one too](https://github.com/bpfkorea/agora/pull/840/files#diff-b11d74d50f129e77728ec32bb7f9b4d9R1367) ?

AndrejMitrovic committed:
Refactor MemBlockStorage

Add ability to set the blockchain which
will be loaded.

Also added a test to ensure load() must
be called before this class is usable.

Removed PreloadedMemBlockStorage.

AndrejMitrovic commented:
Split it into two commits, added a test-case for https://github.com/bpfkorea/agora/commit/c7b90eb9483607d900fef63c3a7b63b2cae2b5dc, and removed `PreloadedMemBlockStorage`.

Geod24 commented on source/agora/node/BlockStorage.d:
```diff
@@ -937,10 +949,6 @@ public class MemBlockStorage : IBlockStorage
         // Basic consistenty checks
         assert(this.height_idx.length == this.hash_idx.length);
         assert(this.height_idx.length == this.blocks.length);
-
-        // Make sure we have no empty block
-        foreach (blk; this.blocks)
-            assert(blk.length > 0);
```
Is this test failing now ?

AndrejMitrovic replied:
Ah I'll revert this, I initially replaced the serialized form with full blocks but that didn't work.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -543,6 +549,17 @@ private mixin template TestNodeMixin ()
 {
     private Registry* registry;
 
+    /// Blocks to preload into the memory storage
+    private immutable(Block)[] blocks;
+
+    ///
+    public this (Config config, Registry* reg, immutable(Block)[] blocks)
+    {
+        this.registry = reg;
+        this.blocks = blocks;
+        super(config);
+    }
+
```
Why aren't the ctor grouped together ?

AndrejMitrovic replied:
Can you explain what you mean?

Geod24 replied:
I need some sleep

Geod24 commented:
Looks roughly good to me. Second commit title looks like a copy-pasta mistake though

AndrejMitrovic commented:
Last commit message is wrong, I'll fix it.

AndrejMitrovic commented:
Fixed commits and reverted other change.
